File: auth/js/src/middleware.js
/**
 * Express middleware for ecosystem authentication.
 * Mirrors the Python FastAPI middleware.
 */

const {
  verifyEcosystemToken,
  verifyRequest,
  NonceStore,
  SIGNATURE_HEADER,
  TIMESTAMP_HEADER,
  NONCE_HEADER,
} = require("./tokens");

// Insecure development default. Mirrors Python's DEFAULT_DEV_SECRET.
const DEFAULT_DEV_SECRET = "dev-ecosystem-secret-change-in-production";

// Process-wide nonce cache for replay detection on the signature path.
const _nonceStore = new NonceStore();

/**
 * Get the shared HMAC secret from environment.
 *
 * Fail-closed everywhere: there is no default. Throws if ECOSYSTEM_HMAC_SECRET
 * is unset or set to the known development default, rather than silently
 * trusting a guessable key.
 * @returns {string}
 */
function getEcosystemSecret() {
  const secret = process.env.ECOSYSTEM_HMAC_SECRET;
  if (!secret) {
    throw new Error(
      "ECOSYSTEM_HMAC_SECRET is not set. A shared secret is required (no default)."
    );
  }
  if (secret === DEFAULT_DEV_SECRET) {
    throw new Error(
      "Refusing the known development default secret. Set a unique ECOSYSTEM_HMAC_SECRET."
    );
  }
  return secret;
}

/**
 * Express middleware that validates ecosystem HMAC-signed requests.
 *
 * Checks X-Ecosystem-Signature header against JSON body,
 * or validates a Bearer token from the Authorization header.
 *
 * On success, sets req.ecosystemAuth with auth details.
 *
 * @param {import('express').Request} req
 * @param {import('express').Response} res
 * @param {import('express').NextFunction} next
 */
function requireEcosystemAuth(req, res, next) {
  const secret = getEcosystemSecret();

  // Check HMAC signature header (replay-resistant request scheme).
  const signature = req.headers[SIGNATURE_HEADER.toLowerCase()];
  if (signature) {
    const timestamp = req.headers[TIMESTAMP_HEADER.toLowerCase()];
    const nonce = req.headers[NONCE_HEADER.toLowerCase()];
    const fullUrl = req.protocol + "://" + req.get("host") + req.originalUrl;

    const isBodyless = req.method === "GET" || req.method === "DELETE";
    const body = isBodyless ? {} : (req.body && typeof req.body === "object" ? req.body : {});

    if (
      !verifyRequest(req.method, fullUrl, secret, signature, timestamp, nonce, body, 300, _nonceStore)
    ) {
      return res
        .status(401)
        .json({ detail: "Invalid, stale, or replayed ecosystem signature" });
    }
    req.ecosystemAuth = { auth_method: "hmac", payload: body };
    return next();
  }

  // Check Bearer token (for service-to-service calls)
  const authHeader = req.headers.authorization;
  if (authHeader && authHeader.startsWith("Bearer ")) {
    const tokenStr = authHeader.slice(7);
    let tokenData;
    try {
      tokenData = JSON.parse(tokenStr);
    } catch {
      return res.status(401).json({ detail: "Invalid token format" });
    }
    if (!verifyEcosystemToken(tokenData, secret)) {
      return res.status(401).json({ detail: "Invalid or expired ecosystem token" });
    }
    req.ecosystemAuth = { auth_method: "token", service: tokenData.service };
    return next();
  }

  return res.status(401).json({ detail: "Missing ecosystem authentication" });
}

module.exports = { requireEcosystemAuth, getEcosystemSecret };

File: auth/python/ecosystem_auth/middleware.py
"""FastAPI middleware/dependency for ecosystem authentication."""

import json
from typing import Callable, Optional

from fastapi import Depends, HTTPException, Request, status
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer

from .tokens import (
    KEY_ID_HEADER,
    NONCE_HEADER,
    NonceStore,
    SIGNATURE_HEADER,
    TIMESTAMP_HEADER,
    get_ecosystem_secret,
    verify_request,
)

security_scheme = HTTPBearer(auto_error=False)

# Process-wide nonce cache for replay detection on the signature path.
_nonce_store = NonceStore()


async def _read_body_obj(request: Request) -> dict:
    raw = await request.body()
    if request.method in ("GET", "DELETE") or not raw:
        return {}
    try:
        return json.loads(raw)
    except Exception:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid JSON body"
        )


async def authenticate_request(
    request: Request,
    credentials: Optional[HTTPAuthorizationCredentials],
    key_resolver: Optional[Callable[[str], Optional[dict]]] = None,
) -> dict:
    """Verify an ecosystem-authenticated request and return its principal.

    Per-app path: when an ``X-Ecosystem-Key-Id`` header is present AND a
    ``key_resolver`` is supplied, resolve the app, verify the HMAC against the
    app's secret, and return that app's principal. Owner path: otherwise verify
    against the shared secret and return the ``__owner__`` principal. Bearer
    tokens are accepted as before. Raises ``HTTPException`` on any failure.
    """
    signature = request.headers.get(SIGNATURE_HEADER)
    if signature:
        timestamp = request.headers.get(TIMESTAMP_HEADER)
        nonce = request.headers.get(NONCE_HEADER)
        body_obj = await _read_body_obj(request)

        key_id = request.headers.get(KEY_ID_HEADER)
        if key_id and key_resolver is not None:
            principal = key_resolver(key_id)
            if not principal:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Unknown or suspended ecosystem key",
                )
            secret = principal["secret"]
            if not verify_request(
                request.method, str(request.url), secret, signature,
                timestamp, nonce, body_obj, nonce_store=_nonce_store,
            ):
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Invalid, stale, or replayed ecosystem signature",
                )
            return {
                "auth_method": "hmac",
                "app_id": principal["app_id"],
                "scopes": list(principal.get("scopes", [])),
                "owned_names": list(principal.get("owned_names", [])),
                "payload": body_obj,
            }

        # Owner path: shared secret.
        secret = get_ecosystem_secret()
        if not verify_request(
            request.method, str(request.url), secret, signature,
            timestamp, nonce, body_obj, nonce_store=_nonce_store,
        ):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid, stale, or replayed ecosystem signature",
            )
        return {
            "auth_method": "hmac",
            "app_id": "__owner__",
            "scopes": ["*"],
            "owned_names": ["*"],
            "payload": body_obj,
        }

    if credentials:
        from .tokens import verify_ecosystem_token

        try:
            token_data = json.loads(credentials.credentials)
        except (json.JSONDecodeError, TypeError):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED, detail="Invalid token format"
            )
        if not verify_ecosystem_token(token_data, get_ecosystem_secret()):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid or expired ecosystem token",
            )
        return {"auth_method": "token", "service": token_data.get("service")}

    raise HTTPException(
        status_code=status.HTTP_401_UNAUTHORIZED,
        detail="Missing ecosystem authentication",
    )


async def require_ecosystem_auth(
    request: Request,
    credentials: Optional[HTTPAuthorizationCredentials] = Depends(security_scheme),
) -> dict:
    """FastAPI dependency: validate an ecosystem-authenticated request (shared-secret/owner)."""
    return await authenticate_request(request, credentials, key_resolver=None)
